Health check

fboucquez opened the pull request:
Hi team, this is the bootstrap health check. It does everything defined in the original issue:

Fixes https://github.com/nemtech/symbol-bootstrap/issues/38

It checks:

- if containers are running
- if the opened ports are listening
- if gateways' /node/health are ok

This health check handles "repeat" and custom ports. 

This can be running with a `symbol-bootstrap run/start --healthCheck` or by calling its own command `symbol-bootstrap healthCheck`

Some other small changes:

- `ipv4_address` and custom `host` alias can be tunned per service (including broker). 
- A docker compose service can be excluded via `excludeDockerService`
- Improved how images are pulled. Better logging and images are prefetched so it doesn't affect compose up booting time 
- I've moved the custom preset examples to its own [file](https://github.com/nemtech/symbol-bootstrap/blob/health-check/docs/presetGuides.md) as I needed to add more. This is part of the doc refactoring @segfaultxavi . 


Fixes https://github.com/nemtech/symbol-bootstrap/issues/30

@Vektrat this will be the [guide](https://github.com/nemtech/symbol-bootstrap/blob/health-check/docs/presetGuides.md#disable-a-service) to disable a service

segfaultxavi commented:
I think `docs/healthCheck.md` is missing from the PR.

Vektrat commented:
thanks ^^

segfaultxavi commented on README.md:
```diff
@@ -75,142 +74,7 @@ Properties in each file override the previous values (by object deep merge).
-- `172.20.0.{{add $index 20}}` will become `172.20.0.30`
-
-The default preset `bootstrap` uses repeat but with just 1 database, 2 peers, 1 api and 1 rest gateway.
+Custom presets gives Symbol Bootstrap its versatility. Checkout the custom preset [guides](docs/presetGuides.md)!
```
```suggestion
Custom presets give Symbol Bootstrap its versatility. Check out the custom preset [guides](docs/presetGuides.md)!
```

segfaultxavi commented on docs/healthCheck.md:
```diff
@@ -0,0 +1,24 @@
+
+## `symbol-bootstrap healthCheck`
+
+It checks if the services created with docker compose are up and running.
```
I would list exactly what things it tests. There's more information in the issue description than in this README...

fboucquez replied:
Fixed. The commands md files are generated from the TS command files.

segfaultxavi commented on docs/run.md:
```diff
@@ -14,21 +14,24 @@ USAGE
+  --resetData          It reset the database and node data but keeps the generated configuration, keys, voting tree
+                       files and block 1
+
+  --timeout=timeout    [default: 60000] If running in detached mode, how long before timing out (in MS)
```
```suggestion
  --timeout=timeout    [default: 60000] If running in detached mode, how long before timing out (in ms)
```
I guess you mean milliseconds, not megasiemens :)

segfaultxavi commented on docs/start.md:
```diff
@@ -29,23 +33,23 @@ OPTIONS
                                           keys, voting tree files and block 1
 
+  --timeout=timeout                       [default: 60000] If running in detached mode, how long before timing out (in
+                                          MS)
```
```suggestion
                                          ms)
```

segfaultxavi reviewed (changes requested):
Besides the small corrections I made, you're mixing different things in the same PR (Addition of `healthCheck` and moving of preset examples to their own README). But hey, it's your repo :)

fboucquez commented:
> Besides the small corrections I made, you're mixing different things in the same PR (Addition of `healthCheck` and moving of preset examples to their own README). But hey, it's your repo :)

I've added more examples and split the file. The docs still need polish, they doc issue will still be open and yours.

segfaultxavi commented on docs/healthCheck.md:
```diff
@@ -1,21 +1,33 @@
 ==============================
 
-It checks if the services created with docker compose are up and running.
+It checks if the services created with docker compose are up and running:
+- if the docker containers are running.
+- if the services' exposed ports are listening.
+- if the rest gateways' /node/health are OK.
```
```suggestion
It checks if the services created with docker compose are up and running.

The performed checks are:
- If the docker containers are running.
- If the services' exposed ports are listening.
- If the rest gateways' /node/health are OK.
```

segfaultxavi commented on docs/healthCheck.md:
```diff
@@ -1,21 +1,33 @@
 ## `symbol-bootstrap healthCheck`
 
-It checks if the services created with docker compose are up and running.
+It checks if the services created with docker compose are up and running:
```
Colon at the end.

fboucquez commented on README.md:
```diff
@@ -398,6 +262,7 @@ General users should install this tool like any other node module.
 * [`symbol-bootstrap clean`](docs/clean.md) - It removes the target folder deleting the generated configuration and data
 * [`symbol-bootstrap compose`](docs/compose.md) - It generates the `docker-compose.yml` file from the configured network.
 * [`symbol-bootstrap config`](docs/config.md) - Command used to set up the configuration files and the nemesis block for the current network
+* [`symbol-bootstrap healthCheck`](docs/healthCheck.md) - It checks if the services created with docker compose are up and running:
```
The actual source is in the `src/commands/healthCheck.ts` file. The cli Oclif framework generates the doc.

segfaultxavi commented on docs/healthCheck.md:
```diff
@@ -0,0 +1,37 @@
+
+It checks if the services created with docker compose are up and running.
+    
+- if the docker containers are running.
+- if the services' exposed ports are listening.
+- if the rest gateways' /node/health are OK.
```
```suggestion
The performed checks are:
- Whether the docker containers are running.
- Whether the services' exposed ports are listening.
- Whether the rest gateways' /node/health are OK.
```

fboucquez replied:
mmm, the "whether" wasn't in the old feedback. 

I cannot commit your change, healthCheck.md is generated. I've changed the typescript source code.

https://github.com/nemtech/symbol-bootstrap/blob/health-check/src/commands/healthCheck.ts#L7

Thanks!

segfaultxavi reviewed (changes requested):
One last minor fix!

Sorry for dragging this for so long, I think some of my previous comments have been lost because I remember I already pointed this one out.

segfaultxavi reviewed (changes requested):
Please be careful. The `if`->`Whether` change is not the only thing I requested.
There is an introductory sentence before the list of checks, otherwise the list does not make sense in the places where it appears on its own.
